test: update test_datasets.py

mattijn opened the pull request:
This PR improves the datasets test suite based on feedback in #3854. Now, the automated tests reflect the implementation, solid work. Namely:
- Fixed test failures: corrected the regex pattern in `test_all_datasets` to properly match error messages (the test was looking for `icon_7zip.png` but the actual error shows `7zip.png`).
- Added data API tests: integrated comprehensive tests for the new data accessor API into the main test file.
- Streamlined test coverage: removed redundant test `is_url` function that validates with vega-datasets 3.2.0 solely precomputed URLs.
- Maintained core functionality: preserved all essential tests for backend functionality, error handling, caching, and data parsing.

The test suite now provides better coverage and the tests are reflecting by their names. No changes to the code in altair were required.

Could you have a look if this addresses your concerns @dangotbanned? 

Close #3854

mattijn committed:
update test_datasets.py

mattijn committed:
fix schema generation

mattijn committed:
Merge branch 'main' into improve-tests-datasets

dangotbanned commented on tools/datasets/datapackage.py:
```diff
@@ -95,7 +95,7 @@ def core(self) -> pl.LazyFrame:
         """Reduce all datasets with schemas to a minimal mapping."""
         m: Any = {
-            Path(rsrc["path"]).stem: {f["name"]: f["type"] for f in s["fields"]}
+            rsrc["name"]: {f["name"]: f["type"] for f in s["fields"]}
```
Ah I remember why I did this now!

For reference, you can see all the schemas like this:

```py
from altair.datasets import Loader


load = Loader.from_backend("polars")
>>> schemas = load._reader._schema_cache.mapping
# very big output
```

But just the names will do for now:

```py
>>> list(schemas)
```

<details><summary><b>Show all dataset file stems with a schema</b></summary>
<p>

```py
['airports',
 'anscombe',
 'barley',
 'birdstrikes',
 'budget',
 'budgets',
 'burtin',
 'cars',
 'co2-concentration',
 'countries',
 'crimea',
 'disasters',
 'driving',
 'flare-dependencies',
 'flare',
 'flights-10k',
 'flights-200k',
 'flights-20k',
 'flights-2k',
 'flights-3m',
 'flights-5k',
 'flights-airport',
 'football',
 'gapminder-health-income',
 'gapminder',
 'github',
 'global-temp',
 'income',
 'iowa-electricity',
 'jobs',
 'la-riots',
 'londonCentroids',
 'lookup_groups',
 'lookup_people',
 'monarchs',
 'movies',
 'normal-2d',
 'obesity',
 'ohlc',
 'penguins',
 'platformer-terrain',
 'political-contributions',
 'population',
 'population_engineers_hurricanes',
 'seattle-weather-hourly-normals',
 'seattle-weather',
 'sp500-2000',
 'sp500',
 'species',
 'stocks',
 'udistrict',
 'unemployment-across-industries',
 'unemployment',
 'uniform-2d',
 'us-employment',
 'us-state-capitals',
 'weather',
 'wheat',
 'windvectors',
 'zipcodes']
```

</p>
</details>

For any of the datasets which have multiple file formats (e.g. `flights-200k.json`, `flights-200k.arrow`) - this let's us use the same schema for both.

IIRC, the `.json` datasets don't get inferred correctly, but in that case the `.arrow` one can be used instead.

On `main`, this works:

```py
from altair.datasets import Loader

load = Loader.from_backend("polars")
>>> load._reader._schema_cache.schema("flights-200k")

Schema([('delay', Int64), ('distance', Int64), ('time', Float64)])
```

mattijn replied:
Yup, but we need the dataset name which is a valid Python identifier and not the url stem

dangotbanned replied:
> Yup, but we need the dataset name which is a valid Python identifier

At the boundary of `data.<...>`? Sure, we need a valid identifer *there*.

But that doesn't mean we **need** to use that name everywhere else - especially if there are benefits to *not* doing that 🙂 

I haven't thought too long on this yet, but considering we have a row like this per-dataset - we *could* do something like adding a `stem` column there - but retrieve it using the snake case name


https://github.com/vega/altair/blob/49a8d770c8aad8d13e932a70002376771164b048/altair/datasets/_typing.py#L191-L203

mattijn replied:
Sure, with the `dataset_name` you can lookup. Isn't it the `file_name` what you are after?

dangotbanned committed:
Merge branch 'main' into improve-tests-datasets

dangotbanned committed:
chore(typing): Ignore `_data.py` warnings

Identifying some of the issues from https://github.com/vega/altair/issues/3854#issuecomment-3073142935

dangotbanned committed:
chore(typing): Ignore `test_datasets.py` warnings

mattijn committed:
stubs to make pylance happy

mattijn committed:
make ruff and mypy happy

mattijn committed:
fix errors generator

mattijn committed:
stubs come from datapackage

mattijn commented:
Not a type expert, let me leave this to you, but the stubs are now derived from the DataPackage and not the other type files.

dangotbanned committed:
revert: back to 49a8d770c8aad8d13e932a70002376771164b048

dangotbanned committed:
fix: Resolve typing issues, remove excessive comments

dangotbanned committed:
revert: Fully restore `test_pandas_date_parse`

https://github.com/vega/altair/issues/3854#issuecomment-3073518839, https://github.com/vega/altair/pull/3857#issuecomment-3077934327

mattijn committed:
bump vega-datasets to v3.2.1

mattijn committed:
more v3.2.1

mattijn committed:
raise for pyarrow backend when date type is in non-ISO date format

mattijn committed:
pass mypy and ruff

mattijn commented:
Thanks for reviewing @dangotbanned! I've included one raise using the `AltairDatasetsError` in case for the PyArrow backend when the date type column is in non-ISO date format, related pyarrow issue: https://github.com/apache/arrow/issues/41488.

This is specifically for the 'stocks' dataset example (date contains values like this:`"Jan 1 2000"`):
```python
from altair.datasets import data

data.stocks(engine="pyarrow")
```

> AltairDatasetsError: PyArrow cannot parse date format in dataset 'stocks'. This is a known limitation of PyArrow's CSV reader for non-ISO date formats.
